mettle/worker/redis.go

```diff
@@ -91,23 +94,33 @@ func (r *redisClient) Healthcheck() error {
 func (r *redisClient) ReadBlob(dg *pb.Digest) ([]byte, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), r.timeout)
 	defer cancel()
-	cmd := r.readRedis.Get(ctx, dg.Hash)
-	if err := cmd.Err(); err == nil {
-		blob, _ := cmd.Bytes()
-		return blob, nil
-	} else if err != redis.Nil {
-		log.Warning("Failed to read blob from Redis: %s", err)
+
+	if r.limiter.Allow() {
+		cmd := r.readRedis.Get(ctx, dg.Hash)
+		if err := cmd.Err(); err == nil {
+			blob, _ := cmd.Bytes()
+			return blob, nil
+		} else if err != redis.Nil {
+			log.Warningf("Failed to read blob from Redis: %s", err)
+			r.limiter.Reserve()
+		}
+	} else {
+		log.Warningf("limiter error rate exceeded, skipping Redis lookup")
 	}
 	// If we get here, the blob didn't exist. Download it then write back to Redis.
 	blob, err := r.elan.ReadBlob(dg)
 	if err != nil {
 		return nil, err
 	}
 	if dg.SizeBytes < r.maxSize {
-		ctx, cancel = context.WithTimeout(context.Background(), r.timeout)
-		defer cancel()
-		if cmd := r.redis.Set(ctx, dg.Hash, blob, 0); cmd.Val() != "OK" {
-			log.Warning("Failed to set blob in Redis: %s", cmd.Err())
+		if r.limiter.Allow() {
```

Comment:
Not sure whether this needs to be in the rate limit given that I've moved this to be in a go routine anyway. Feels that we want to skip _all_ Redis or not in one single movement.

I suppose this might be clearer with a top level `if allow` that delegates to the Elan client and skips everything?

Reply:
It's probably good not to spin off a bunch of goroutines if we think Redis is down anyway?

Reply:
I've ended up removing the limiter on writes, so we keep attmping to upload to Redis, even if the reads are failing. 

The timeout is short enough at 1 second that I think this is _probably_ fine from a memory perspective.